File: 2024/day2/solution.go
package main

import (
	adventofcode "adventofcode/utils"
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {
	reports := getReportsFromInput("input")

	numberOfSafeReports := 0
	for _, report := range reports {
		if reportIsSafe(report) {
			numberOfSafeReports += 1
		}
	}

	fmt.Println("Number of safe reports (part 1): ", numberOfSafeReports)

	numberOfCorrectedSafeReports := 0

	for _, report := range reports {
		if !reportIsSafe(report) {
			for i := 0; i < len(report); i++ {

				// create new slice and copy all elements except the one at index i
				newReport := make([]int, len(report)-1)
				copy(newReport, report[:i])
				copy(newReport[i:], report[i+1:])

				if reportIsSafe(newReport) {
					numberOfCorrectedSafeReports += 1
					break
				}
			}
		} else {
			numberOfCorrectedSafeReports += 1
		}
	}

	fmt.Println("Number of corrected safe reports (part 2): ", numberOfCorrectedSafeReports)
}

func allAscending(report []int) bool {
	for i := 1; i < len(report); i++ {
		if report[i-1] >= report[i] {
			return false
		}
	}
	return true
}

func allDescending(report []int) bool {
	for i := 1; i < len(report); i++ {
		if report[i-1] <= report[i] {
			return false
		}
	}
	return true
}

func reportIsSafe(report []int) bool {
	if !allAscending(report) && !allDescending(report) {
		return false
	}

	for i := 0; i < len(report)-1; i++ {
		if math.Abs(float64(report[i])-float64(report[i+1])) > 3 {
			return false
		}
	}

	return true
}

func getReportsFromInput(in string) [][]int {
	content := adventofcode.GetFileContent(in)

	reports := [][]int{}

	scanner := bufio.NewScanner(strings.NewReader(content))
	for scanner.Scan() {
		line := scanner.Text()
		report := []int{}
		for _, id := range strings.Split(line, " ") {
			convertedID, err := strconv.Atoi(id)
			if err == nil {
				report = append(report, convertedID)
			}
		}
		reports = append(reports, report)
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, "reading input:", err)
	}

	return reports
}

File: 2024/day2/solution.py
from typing import List

from utils import pyutils as adventutils


def formatContent(content: List[str]) -> List[List[int]]:
    return [list(map(int, line.split(" "))) for line in content]


content = formatContent(adventutils.file_contest_as_list_of_lines("./day2/input"))


def reportIsSafe(report: List[int]) -> bool:
    ascending = all(report[i] < report[i + 1] for i in range(len(report) - 1))
    descending = all(report[i] > report[i + 1] for i in range(len(report) - 1))

    if not (ascending or descending):
        return False

    for i in range(len(report) - 1):
        if abs(report[i] - report[i + 1]) > 3:
            return False

    return True


def part1(reports: List[str]) -> int:
    return [reportIsSafe(report) for report in reports].count(True)


def part2(reports: List[str]) -> int:
    numberOfSafeReports = 0
    for report in reports:
        if not reportIsSafe(report):
            monte_carlo_reports = []
            # RIP memory (TODO: improve this)
            # create one list for each item in the report with that item missing
            for i in range(len(report)):
                new_list = report[:i] + report[i + 1 :]
                monte_carlo_reports.append(new_list)

            for shortened_report in monte_carlo_reports:
                if reportIsSafe(shortened_report):
                    numberOfSafeReports += 1
                    break
        else:
            numberOfSafeReports += 1

    return numberOfSafeReports


print("number of safe reports (part 1): ", part1(content))
print("number of safe reports with correction (part2)", part2(content))
